rotkehlchen/chain/ethereum/balancer/balancer.py

```diff
@@ -0,0 +1,321 @@
+        for pool_share in result['poolShares']:
+            # ! Checksum address
+            address = to_checksum_address(pool_share['userAddress']['id'])
+            user_pool_balance = FVal(pool_share['balance'])
+            pool = pool_share['poolId']
+            pool_balance = FVal(pool['totalShares'])
+
+            bpool_assets = []
+            for token in pool['tokens']:
+                # Try to get the token <Asset> and classify as known or unknown
+                # ! Checksum address
+                token_address = to_checksum_address(token['address'])
+                token_symbol = token['symbol']
+                asset = None
+                try:
+                    # ! TODO VN PR: need optimisations? symbol uniqueness?
+                    asset = Asset(token_symbol)
+                except UnknownAsset:
+                    unknown_assets.add(token_address)
+                    log.error(f'Encountered unknown asset {token_symbol} in balancer. Skipping')
+                else:
+                    known_assets.add(KnownAsset(address=token_address, asset=asset))
+
+                # Estimate underlying asset balance given user pool balance
+                asset_balance = FVal(token['balance'])
+                user_asset_balance = (user_pool_balance / pool_balance * asset_balance)
+
+                bpool_asset = BalancerPoolAsset(
+                    address=token_address,
+                    balance=asset_balance,
+                    denorm_weight=FVal(token['denormWeight']),
+                    name=token['name'],
+                    symbol=token_symbol,
+                    user_balance=user_asset_balance,
+                    asset=asset,
+                )
+                bpool_assets.append(bpool_asset)
+
+            bpool = BalancerPool(
+                address=to_checksum_address(pool['id']),  # ! Checksum address
+                assets=bpool_assets,
+                assets_count=FVal(pool['tokensCount']),
+                symbol=pool['symbol'],
+                balance=pool_balance,
+                weight=FVal(pool['totalWeight']),
+                user_balance=user_pool_balance,
+            )
+            addresses_balancer_pools[address].append(bpool)
+
+        return BalancerBalances(
+            addresses_balancer_pools=addresses_balancer_pools,
+            known_assets=known_assets,
+            unknown_assets=unknown_assets,
+        )
+
+    def _get_known_assets_prices(
+        self,
+        known_assets: Set[KnownAsset],
+        unknown_assets: Set[ChecksumEthAddress],
+    ) -> AssetsPrices:
+        """Get the tokens prices via Inquirer
+
+        Given an asset, if `find_usd_price()` returns zero, it will be added
+        into `unknown_assets`.
+        """
+        known_asset_prices: AssetsPrices = {}
+        known_assets_ = list(known_assets)  # ! Get a deterministic sequence
+        # ! TODO VN PR: consider timeout and list/kill greenlets
+        assets_usd_prices = [
```

Comment:
I am not a fan of the naming here. What this loop comprehension returns is a "job", not the result itself. You then need to query its `value`. You do it below, but the naming may lead to confusion to someone who does not know how gevent works.

Reply:
I see. To be honest is my first experience with `gevent` and `greenlets`. I never had the opportunity to use them, and this is why I played safe at the beginning (and the amount of questions).

Reply:
Rolled back to sequential requests at the moment.